src/seq.rs

```diff
@@ -475,49 +480,97 @@ impl PortSubscribe {
 
 }
 
+#[derive(Clone)]
+enum EventExtraData<'a> {
+    None,
+    Borrowed(&'a [u8]),
+    Owned(Vec<u8>)
+}
+
+impl<'a> EventExtraData<'a> {
+    fn into_owned<'b>(self) -> EventExtraData<'b> {
+        match self {
+            EventExtraData::None => EventExtraData::None,
+            EventExtraData::Borrowed(buf) => EventExtraData::Owned(buf.to_vec()),
+            EventExtraData::Owned(vec) => EventExtraData::Owned(vec)
+        }
+    }
+}
+
 /// [snd_seq_event_t](http://www.alsa-project.org/alsa-doc/alsa-lib/structsnd__seq__event__t.html) wrapper
 ///
 /// Fields of the event is not directly exposed. Instead call `Event::new` to set data (which can be, e g, an EvNote).
 /// Use `get_type` and `get_data` to retreive data.
-pub struct Event(alsa::snd_seq_event_t, EventType, Option<Vec<u8>>);
+///
+/// The lifetime parameter refers to the lifetime of an associated external buffer that might be used for
+/// variable-length messages (e.g. SysEx).
+pub struct Event<'a>(alsa::snd_seq_event_t, EventType, EventExtraData<'a>);
 
-unsafe impl Send for Event {}
+unsafe impl<'a> Send for Event<'a> {}
 
-impl Event {
-    pub fn new<D: EventData>(t: EventType, data: &D) -> Self {
-        let mut z = Event(unsafe { mem::zeroed() }, EventType::None, None);
+impl<'a> Event<'a> {
+    pub fn new<D: SetEventData>(t: EventType, data: D) -> D::Output {
+        let mut z = Event(unsafe { mem::zeroed() }, EventType::None, EventExtraData::None);
         z.1 = t;
         (z.0)._type = t as c_uchar;
+        let length_flags = match t {
```

Comment:
I did it in `new` because the length flags only depend on the event type, and `Event` has no method to change the type, so it's not necessary to "fix" it whenever `ensure_buf` is called. Events that are returned from ALSA should always have the correct flags set, so it's only a problem when they are created manually using `Event::new`.